ztest/ztest.go

```diff
@@ -271,15 +374,180 @@ func FromYAMLFile(filename string) (*ZTest, error) {
 	return &z, nil
 }
 
-// Run runs the query in ZQL over inputs and returns the output formatted
+func diffErr(expected, actual string) error {
+	diff, _ := difflib.GetUnifiedDiffString(difflib.UnifiedDiff{
+		A:        difflib.SplitLines(expected),
+		FromFile: "expected",
+		B:        difflib.SplitLines(actual),
+		ToFile:   "actual",
+		Context:  5,
+	})
+	return fmt.Errorf("expected and actual outputs differ:\n%s", diff)
+}
+
+func run(t *testing.T, testname, bindir, dirname, filename string, zt *ZTest) {
+	if err := zt.check(); err != nil {
+		t.Fatalf("%s: bad yaml format: %s", filename, err)
+	}
+	if zt.Script != "" {
+		if bindir == "" {
+			t.Skip("skipping script test on in-process run")
+		}
+		adir, _ := filepath.Abs(dirname)
+		err := runsh(testname, bindir, adir, zt)
+		if err != nil {
+			t.Fatalf("%s: %s", filename, err)
+		}
+		return
+	}
+	out, errout, err := runzq(bindir, zt.ZQL, zt.OutputFormat, zt.OutputFlags, zt.Input...)
+	if err != nil {
+		if zt.errRegex != nil {
+			if !zt.errRegex.Match([]byte(errout)) {
+				t.Fatalf("%s: error doesn't match expected error regex: %s %s", filename, zt.ErrorRE, errout)
+			}
+		} else {
+			if out != "" {
+				out = "\noutput:\n" + out
+			}
+			t.Fatalf("%s: %s%s", filename, err, out)
+		}
+	} else if zt.errRegex != nil {
+		t.Fatalf("%s: no error when expecting error regex: %s", filename, zt.ErrorRE)
+	}
+	expectedOut, oerr := zt.getOutput()
+	require.NoError(t, oerr)
+	if out != expectedOut {
+		a := expectedOut
+		b := out
+
+		if !utf8.ValidString(a) {
+			a = encodeHex(a)
+			b = encodeHex(b)
+		}
+
+		err := diffErr(a, b)
+		t.Fatalf("%s: expected and actual outputs differ:\n%s", filename, err)
+	}
+	if err == nil && errout != zt.Warnings {
+		err := diffErr(zt.Warnings, errout)
+		t.Fatalf("%s: expected and actual warnings differ:\n%s", filename, err)
+	}
+}
+
+func checkPatterns(patterns map[string]*regexp.Regexp, dir *Dir, stdout, stderr string) error {
+	for name, re := range patterns {
+		var body []byte
+		switch name {
+		case "stdout":
+			body = []byte(stdout)
+		case "stderr":
+			body = []byte(stderr)
+		default:
+			var err error
+			body, err = dir.Read(name)
+			if err != nil {
+				return fmt.Errorf("%s: %s", name, err)
+			}
+		}
+		if !re.Match(body) {
+			return fmt.Errorf("regex mismatch: %s %s", re, string(body))
+		}
+	}
+	return nil
+}
+
+func checkData(files map[string][]byte, dir *Dir, stdout, stderr string) error {
+	for name, expected := range files {
+		var actual []byte
+		switch name {
+		case "stdout":
+			actual = []byte(stdout)
+		case "stderr":
+			actual = []byte(stderr)
+		default:
+			var err error
+			actual, err = dir.Read(name)
+			if err != nil {
+				return fmt.Errorf("%s: %s", name, err)
+			}
+		}
+		if !bytes.Equal(expected, actual) {
+			return diffErr(string(expected), string(actual))
+		}
+	}
+	return nil
+}
+
+func runsh(testname, bindir, dirname string, zt *ZTest) error {
+	dir, err := NewDir(testname, dirname)
```

Comment:
This creates a temporary directory in dirname, which is inside the source tree. Is that deliberate? (Our other tests do ioutil.TempDir("", pattern) to use the default location for temporary directories.)

Reply:
fix has been pushed (this was from old logic that would leave the testdir around if the test failed for debugging)